<a href="https://colab.research.google.com/github/briandankowski/GuidedCapstone/blob/master/Final_GuidedCapstone_Step4_and_Step5HL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
os.listdir()

['.config', 'sample_data']

In [3]:
!mkdir data
!mkdir figures
!mkdir models
from google.colab import files
files.upload()

Saving step3_output.csv to step3_output.csv


{'step3_output.csv': b"Name,state,summit_elev,vertical_drop,fastQuads,quad,triple,double,surface,total_chairs,Runs,TerrainParks,LongestRun_mi,SkiableTerrain_ac,Snow Making_ac,daysOpenLastYear,yearsOpen,averageSnowfall,AdultWeekday,AdultWeekend,projectedDaysOpen,NightSkiing_ac,clusters\nAlyeska Resort,Alaska,3939,2500,2,2,0,0,2,7,76.0,2.0,1.0,1610.0,113.0,150.0,60.0,669.0,65.0,85.0,150.0,550.0,1\nEaglecrest Ski Area,Alaska,2600,1540,0,0,0,4,0,4,36.0,1.0,2.0,640.0,60.0,45.0,44.0,350.0,47.0,53.0,90.0,0.0,1\nHilltop Ski Area,Alaska,2090,294,0,0,1,0,2,3,13.0,1.0,1.0,30.0,30.0,150.0,36.0,69.0,30.0,34.0,152.0,30.0,1\nArizona Snowbowl,Arizona,11500,2300,0,2,2,1,2,8,55.0,4.0,2.0,777.0,104.0,122.0,81.0,260.0,89.0,89.0,122.0,0.0,0\nSunrise Park Resort,Arizona,11100,1800,1,2,3,1,0,7,65.0,2.0,1.2,800.0,80.0,115.0,49.0,250.0,74.0,78.0,104.0,80.0,0\nYosemite Ski & Snowboard Area,California,7800,600,0,0,1,3,1,5,10.0,2.0,0.4,88.0,0.0,110.0,84.0,300.0,47.0,47.0,107.0,0.0,3\nBear Mountain,California,8805

In [4]:
os.listdir()

['.config', 'figures', 'data', 'models', 'step3_output.csv', 'sample_data']

In [5]:
!rm 'step3_output (1).csv'

rm: cannot remove 'step3_output (1).csv': No such file or directory


In [6]:
os.listdir()

['.config', 'figures', 'data', 'models', 'step3_output.csv', 'sample_data']

In [7]:
os.getcwd()

'/content'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [8]:
df = pd.read_csv('step3_output.csv', index_col= None)
df.describe().T

count         mean  ...          75%      max
summit_elev        330.0  4591.818182  ...  7806.000000  13487.0
vertical_drop      330.0  1215.427273  ...  1800.000000   4425.0
fastQuads          330.0     1.018182  ...     1.000000     15.0
quad               330.0     0.933333  ...     1.000000      8.0
triple             330.0     1.500000  ...     2.000000      8.0
double             330.0     1.833333  ...     3.000000     14.0
surface            330.0     2.621212  ...     3.000000     15.0
total_chairs       330.0     8.266667  ...    10.000000     41.0
Runs               330.0    48.214724  ...    60.000000    341.0
TerrainParks       330.0     2.384848  ...     3.000000     14.0
LongestRun_mi      330.0     1.433231  ...     2.000000      6.0
SkiableTerrain_ac  330.0   739.801223  ...   739.801223  26819.0
Snow Making_ac     330.0   150.496970  ...   192.000000   3379.0
daysOpenLastYear   330.0   115.103943  ...   130.000000    305.0
yearsOpen          330.0    63.656535  ...    68.750000   2019.0
averageSnowfall    330.0   185.316456  ...   278.750000    669.0
AdultWeekday       330.0    57.916957  ...    67.000000    179.0
AdultWeekend       330.0    64.166810  ...    72.000000    179.0
projectedDaysOpen  330.0   120.053004  ...   136.000000    305.0
NightSkiing_ac     330.0    56.890909  ...    83.750000    650.0
clusters           330.0     1.306061  ...     1.000000      3.0

[21 rows x 8 columns]

In [9]:
df.shape

(330, 23)

In [10]:
df.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         1
1  Eaglecrest Ski Area   Alaska  ...             0.0         1
2     Hilltop Ski Area   Alaska  ...            30.0         1
3     Arizona Snowbowl  Arizona  ...             0.0         0
4  Sunrise Park Resort  Arizona  ...            80.0         0

[5 rows x 23 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [11]:
# select object type columns
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(df['state'])], axis=1)
df.head()

Name  summit_elev  ...  Wisconsin  Wyoming
0       Alyeska Resort         3939  ...          0        0
1  Eaglecrest Ski Area         2600  ...          0        0
2     Hilltop Ski Area         2090  ...          0        0
3     Arizona Snowbowl        11500  ...          0        0
4  Sunrise Park Resort        11100  ...          0        0

[5 rows x 57 columns]

## Standardize the magnitude of numeric features
**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [12]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called x,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df.drop(['Name','AdultWeekend'], axis=1)
print(x)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter x 
x_scaled=scaler.transform(x)

     summit_elev  vertical_drop  fastQuads  ...  West Virginia  Wisconsin  Wyoming
0           3939           2500          2  ...              0          0        0
1           2600           1540          0  ...              0          0        0
2           2090            294          0  ...              0          0        0
3          11500           2300          0  ...              0          0        0
4          11100           1800          1  ...              0          0        0
..           ...            ...        ...  ...            ...        ...      ...
325         9500           1000          0  ...              0          0        1
326         7428            810          0  ...              0          0        1
327         7808           1571          0  ...              0          0        1
328         9663            990          0  ...              0          0        1
329         9500           1100          0  ...              0          0        1

[33

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [13]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to x_scaled and y 
# Declare four variables, s_train, s_test, y_train and y_test separated by commas 
s_train, s_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.25, random_state=1)

In [14]:
y

array([ 85.        ,  53.        ,  34.        ,  89.        ,
        78.        ,  47.        ,  64.16681004,  64.16681004,
        64.16681004,  78.        ,  75.        ,  64.16681004,
        64.16681004,  64.16681004, 159.        ,  59.        ,
        84.        ,  69.        ,  64.16681004,  64.16681004,
        83.        ,  64.16681004,  79.        ,  50.        ,
       125.        ,  69.        ,  85.        , 179.        ,
        64.16681004,  64.16681004, 158.        ,  64.16681004,
        89.        ,  64.16681004,  25.        ,  64.16681004,
        79.        ,  89.        ,  71.        ,  79.        ,
        56.        ,  84.        ,  64.16681004,  65.        ,
       139.        ,  64.16681004,  64.16681004,  72.        ,
        65.        ,  60.        ,  55.        ,  62.        ,
        42.        ,  64.        ,  70.        ,  42.        ,
        47.        ,  32.        ,  47.        ,  64.16681004,
        81.        ,  62.        ,  43.        ,  64.16

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

# **Model 1**

In [15]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(s_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [16]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(s_test)

In [17]:
y_pred

array([ 58.31135206,  89.46955518,  78.41877393,  75.67072706,
        70.57111768,  76.15510206,  44.82893018,  46.79377393,
        93.51057081,  84.89533643,  50.45197706,  79.25080518,
        50.93244581,  70.93830518,  62.45978956,  93.39533643,
        42.60822706,  72.86994581,  61.73127393,  40.74885206,
       113.47346143,  36.96955518,  53.65314893,  49.23322706,
        48.09260206,  62.58478956,  75.87482862,  74.68439893,
        64.26643018,  66.54572706,  35.05744581,  58.87189893,
        80.33674268,  57.13947706,  75.38947706,  49.37385206,
        22.96760206,  29.40510206,  61.64728956,  37.28010206,
        48.50080518,  63.23810987,  86.71369581,  72.23713331,
        72.41096143,  59.18439893,  59.49299268, 166.31135206,
        37.67072706,  61.15998487,  68.18439893,  61.17463331,
        53.16486768,  44.33088331,  59.32795362,  43.11799268,
        84.96076612,  39.33478956, 120.62385206, 119.29963331,
        55.93635206,  53.96857862,  41.79572706,  81.09

## Review Model Outcomes — Iterate over additional models as needed

In [18]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [19]:
explained_variance_score(y_true=y_test, y_pred=y_pred)

0.9409335239835701

In [20]:
mean_absolute_error(y_true=y_test, y_pred=y_pred)

4.915690182064833

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [21]:
lm.intercept_

64.11701612065369

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [22]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data = []
coef_data = pd.DataFrame(abs(lm.coef_), x.columns, columns=['Coefficient'])
print(coef_data)


                    Coefficient
summit_elev        1.899311e+00
vertical_drop      9.583299e-01
fastQuads          1.157118e+00
quad               1.527711e+00
triple             2.236678e+00
double             1.312403e+00
surface            3.399401e-01
total_chairs       1.120816e+00
Runs               9.005197e-01
TerrainParks       9.941889e-02
LongestRun_mi      4.374870e-01
SkiableTerrain_ac  2.009429e-01
Snow Making_ac     8.709411e-01
daysOpenLastYear   5.413062e-01
yearsOpen          3.547436e-01
averageSnowfall    2.204681e-02
AdultWeekday       1.923904e+01
projectedDaysOpen  1.469920e-01
NightSkiing_ac     6.434278e-01
clusters           4.043513e-01
Alaska             2.861859e+12
Arizona            2.340268e+12
California         7.360420e+12
Colorado           7.521430e+12
Connecticut        3.683328e+12
Idaho              5.644402e+12
Illinois           3.299533e+12
Indiana            2.340268e+12
Iowa               2.861859e+12
Maine              4.911199e+12
Maryland

In [23]:
coef_data.sort_values(['Coefficient'], ascending=False)

Coefficient
New York           9.045841e+12
Michigan           8.536817e+12
Colorado           7.521430e+12
California         7.360420e+12
Pennsylvania       7.023777e+12
New Hampshire      6.476473e+12
Wisconsin          6.476473e+12
Vermont            6.280795e+12
Minnesota          6.077449e+12
Utah               5.865635e+12
Idaho              5.644402e+12
Montana            5.644402e+12
Massachusetts      5.412594e+12
Washington         5.168788e+12
Oregon             5.168788e+12
Maine              4.911199e+12
New Mexico         4.911199e+12
Wyoming            4.637529e+12
North Carolina     4.028672e+12
Connecticut        3.683328e+12
Ohio               3.683328e+12
West Virginia      3.299533e+12
Virginia           3.299533e+12
Illinois           3.299533e+12
Nevada             3.299533e+12
Alaska             2.861859e+12
Iowa               2.861859e+12
New Jersey         2.340268e+12
Missouri           2.340268e+12
Indiana            2.340268e+12
Arizona            2.340268e+12
South Dakota       2.340268e+12
Tennessee          1.657340e+12
Maryland           1.657340e+12
Rhode Island       1.657340e+12
AdultWeekday       1.923904e+01
triple             2.236678e+00
summit_elev        1.899311e+00
quad               1.527711e+00
double             1.312403e+00
fastQuads          1.157118e+00
total_chairs       1.120816e+00
vertical_drop      9.583299e-01
Runs               9.005197e-01
Snow Making_ac     8.709411e-01
NightSkiing_ac     6.434278e-01
daysOpenLastYear   5.413062e-01
LongestRun_mi      4.374870e-01
clusters           4.043513e-01
yearsOpen          3.547436e-01
surface            3.399401e-01
SkiableTerrain_ac  2.009429e-01
projectedDaysOpen  1.469920e-01
TerrainParks       9.941889e-02
averageSnowfall    2.204681e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

# Model 2 

In [24]:
df2 = pd.read_csv('step3_output.csv', index_col= None)

In [25]:
df2.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         1
1  Eaglecrest Ski Area   Alaska  ...             0.0         1
2     Hilltop Ski Area   Alaska  ...            30.0         1
3     Arizona Snowbowl  Arizona  ...             0.0         0
4  Sunrise Park Resort  Arizona  ...            80.0         0

[5 rows x 23 columns]

In [26]:
# Declare an explanatory variable, called x2, and assign it the result of dropping 'state' 'Name' and 'AdultWeekend' from the df2
x2 = df2.drop(['state','Name','AdultWeekend'], axis=1)
print(x2)

# Declare a response variable, called y2, and assign it the AdultWeekend column of the df2 
y2 = df2['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x 
scaler2 = preprocessing.StandardScaler().fit(x2)

# Declare a variable called x2_scaled, and assign it the result of calling the transform() method with parameter x2 
x2_scaled=scaler2.transform(x2)

     summit_elev  vertical_drop  ...  NightSkiing_ac  clusters
0           3939           2500  ...           550.0         1
1           2600           1540  ...             0.0         1
2           2090            294  ...            30.0         1
3          11500           2300  ...             0.0         0
4          11100           1800  ...            80.0         0
..           ...            ...  ...             ...       ...
325         9500           1000  ...             0.0         3
326         7428            810  ...             0.0         3
327         7808           1571  ...           110.0         3
328         9663            990  ...             0.0         3
329         9500           1100  ...             0.0         3

[330 rows x 20 columns]


In [27]:
# Get the 1-dimensional flattened array of our response variable y2 by calling the ravel() function on y2
y2 = y2.ravel()
# Call the train_test_split() function with the first two parameters set to x2_scaled and y2 
# Declare four variables, s2_train, s2_test, y2_train and y2_test separated by commas 
s2_train, s2_test, y2_train, y2_test = train_test_split(x2_scaled, y2, test_size=0.25, random_state=1)

In [28]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(s2_train,y2_train)

In [29]:
# Make a variable called y2_pred and assign it the result of calling predict() on our model variable with parameter x2_test
y2_pred = model2.predict(s2_test)

In [30]:
explained_variance_score(y_true=y2_test, y_pred=y2_pred)

0.9296939181024035

In [31]:
mean_absolute_error(y_true=y2_test, y_pred=y2_pred)

5.3383259592237975

In [32]:
lm2.intercept_

64.07539503650082

In [33]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data2 = []
coef_data2 = pd.DataFrame(abs(lm2.coef_), x2.columns, columns=['Coefficient'])
print(coef_data2)

                   Coefficient
summit_elev           0.994699
vertical_drop         1.527316
fastQuads             0.223443
quad                  1.243351
triple                1.112562
double                0.034431
surface               1.520371
total_chairs          1.575504
Runs                  1.235018
TerrainParks          0.130860
LongestRun_mi         0.289668
SkiableTerrain_ac     0.253621
Snow Making_ac        0.432130
daysOpenLastYear      1.104171
yearsOpen             0.269139
averageSnowfall       1.691297
AdultWeekday         20.080300
projectedDaysOpen     0.331176
NightSkiing_ac        0.426303
clusters              0.271374


In [34]:
coef_data2.sort_values(['Coefficient'], ascending=False)

Coefficient
AdultWeekday         20.080300
averageSnowfall       1.691297
total_chairs          1.575504
vertical_drop         1.527316
surface               1.520371
quad                  1.243351
Runs                  1.235018
triple                1.112562
daysOpenLastYear      1.104171
summit_elev           0.994699
Snow Making_ac        0.432130
NightSkiing_ac        0.426303
projectedDaysOpen     0.331176
LongestRun_mi         0.289668
clusters              0.271374
yearsOpen             0.269139
SkiableTerrain_ac     0.253621
fastQuads             0.223443
TerrainParks          0.130860
double                0.034431

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

# **Model 3**

In [35]:
df3 = pd.read_csv('step3_output.csv', index_col= None)

In [36]:
df3.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         1
1  Eaglecrest Ski Area   Alaska  ...             0.0         1
2     Hilltop Ski Area   Alaska  ...            30.0         1
3     Arizona Snowbowl  Arizona  ...             0.0         0
4  Sunrise Park Resort  Arizona  ...            80.0         0

[5 rows x 23 columns]

In [37]:
# Declare an explanatory variable, called x3, and assign it the result of dropping 'state', 'Name','summit_elev' and 'AdultWeekend' from the df3
x3 = df3.drop(['state','Name','AdultWeekend','summit_elev'], axis=1)
print(x3)

# Declare a response variable, called y3, and assign it the AdultWeekend column of the df3 
y3 = df3['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x3 
scaler3 = preprocessing.StandardScaler().fit(x3)

# Declare a variable called x3_scaled, and assign it the result of calling the transform() method with parameter x3 
x3_scaled=scaler3.transform(x3)

     vertical_drop  fastQuads  ...  NightSkiing_ac  clusters
0             2500          2  ...           550.0         1
1             1540          0  ...             0.0         1
2              294          0  ...            30.0         1
3             2300          0  ...             0.0         0
4             1800          1  ...            80.0         0
..             ...        ...  ...             ...       ...
325           1000          0  ...             0.0         3
326            810          0  ...             0.0         3
327           1571          0  ...           110.0         3
328            990          0  ...             0.0         3
329           1100          0  ...             0.0         3

[330 rows x 19 columns]


In [38]:
# Get the 1-dimensional flattened array of our response variable y3 by calling the ravel() function on y3
y3 = y3.ravel()
# Call the train_test_split() function with the first two parameters set to x3_scaled and y3 
# Declare four variables, s3_train, s3_test, y3_train and y3_test separated by commas 
s3_train, s3_test, y3_train, y3_test = train_test_split(x3_scaled, y3, test_size=0.25, random_state=1)

In [39]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(s3_train,y3_train)

In [40]:
# Make a variable called y3_pred and assign it the result of calling predict() on our model variable with parameter x3_test
y3_pred = model3.predict(s3_test)

In [41]:
explained_variance_score(y_true=y3_test, y_pred=y3_pred)

0.9263168900468063

In [42]:
mean_absolute_error(y_true=y3_test, y_pred=y3_pred)

5.471497752071906

In [43]:
lm3.intercept_

64.07436966949007

In [44]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data3 = []
coef_data3 = pd.DataFrame(abs(lm3.coef_), x3.columns, columns=['Coefficient'])
print(coef_data3)

                   Coefficient
vertical_drop         1.090355
fastQuads             0.299072
quad                  1.329484
triple                1.105076
double                0.052905
surface               1.536132
total_chairs          1.592218
Runs                  1.310954
TerrainParks          0.107832
LongestRun_mi         0.272616
SkiableTerrain_ac     0.319464
Snow Making_ac        0.395067
daysOpenLastYear      0.922663
yearsOpen             0.292590
averageSnowfall       2.066468
AdultWeekday         20.014486
projectedDaysOpen     0.139252
NightSkiing_ac        0.530249
clusters              0.321841


In [45]:
coef_data3.sort_values(['Coefficient'], ascending=False)

Coefficient
AdultWeekday         20.014486
averageSnowfall       2.066468
total_chairs          1.592218
surface               1.536132
quad                  1.329484
Runs                  1.310954
triple                1.105076
vertical_drop         1.090355
daysOpenLastYear      0.922663
NightSkiing_ac        0.530249
Snow Making_ac        0.395067
clusters              0.321841
SkiableTerrain_ac     0.319464
fastQuads             0.299072
yearsOpen             0.292590
LongestRun_mi         0.272616
projectedDaysOpen     0.139252
TerrainParks          0.107832
double                0.052905

# **Model 4 - No Clusters Column**

In [46]:
df4 = pd.read_csv('step3_output.csv', index_col= None)

In [47]:
df4.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         1
1  Eaglecrest Ski Area   Alaska  ...             0.0         1
2     Hilltop Ski Area   Alaska  ...            30.0         1
3     Arizona Snowbowl  Arizona  ...             0.0         0
4  Sunrise Park Resort  Arizona  ...            80.0         0

[5 rows x 23 columns]

In [48]:
# Declare an explanatory variable, called x4, and assign it the result of dropping 'clusters','state','Name','summit_elev' and 'AdultWeekend' from the df4
x4 = df4.drop(['state','Name','AdultWeekend','summit_elev','clusters'], axis=1)
print(x4)

# Declare a response variable, called y4, and assign it the AdultWeekend column of the df4 
y4 = df4['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x4 
scaler4 = preprocessing.StandardScaler().fit(x4)

# Declare a variable called x4_scaled, and assign it the result of calling the transform() method with parameter x4 
x4_scaled=scaler4.transform(x4)

     vertical_drop  fastQuads  ...  projectedDaysOpen  NightSkiing_ac
0             2500          2  ...         150.000000           550.0
1             1540          0  ...          90.000000             0.0
2              294          0  ...         152.000000            30.0
3             2300          0  ...         122.000000             0.0
4             1800          1  ...         104.000000            80.0
..             ...        ...  ...                ...             ...
325           1000          0  ...         120.053004             0.0
326            810          0  ...          77.000000             0.0
327           1571          0  ...         123.000000           110.0
328            990          0  ...         120.053004             0.0
329           1100          0  ...         120.053004             0.0

[330 rows x 18 columns]


In [49]:
# Get the 1-dimensional flattened array of our response variable y4 by calling the ravel() function on y4
y4 = y4.ravel()
# Call the train_test_split() function with the first two parameters set to x4_scaled and y4 
# Declare four variables, s4_train, s4_test, y4_train and y4_test separated by commas 
s4_train, s4_test, y4_train, y4_test = train_test_split(x4_scaled, y4, test_size=0.25, random_state=1)

In [50]:
lm4 = linear_model.LinearRegression()
model4 = lm4.fit(s4_train,y4_train)

In [51]:
# Make a variable called y4_pred and assign it the result of calling predict() on our model variable with parameter x4_test
y4_pred = model4.predict(s4_test)

In [52]:
explained_variance_score(y_true=y4_test, y_pred=y4_pred)

0.9248663129115536

In [53]:
mean_absolute_error(y_true=y4_test, y_pred=y4_pred)

5.542609970084944

In [54]:
lm4.intercept_

64.07178997728322

In [55]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data4 = []
coef_data4 = pd.DataFrame(abs(lm4.coef_), x4.columns, columns=['Coefficient'])
print(coef_data4)

                   Coefficient
vertical_drop         1.011005
fastQuads             0.345907
quad                  1.336541
triple                1.107624
double                0.048383
surface               1.491233
total_chairs          1.564824
Runs                  1.190480
TerrainParks          0.110023
LongestRun_mi         0.291573
SkiableTerrain_ac     0.314681
Snow Making_ac        0.447529
daysOpenLastYear      0.976423
yearsOpen             0.294256
averageSnowfall       2.163483
AdultWeekday         20.068013
projectedDaysOpen     0.184608
NightSkiing_ac        0.474297


In [56]:
coef_data4.sort_values(['Coefficient'], ascending=False)

Coefficient
AdultWeekday         20.068013
averageSnowfall       2.163483
total_chairs          1.564824
surface               1.491233
quad                  1.336541
Runs                  1.190480
triple                1.107624
vertical_drop         1.011005
daysOpenLastYear      0.976423
NightSkiing_ac        0.474297
Snow Making_ac        0.447529
fastQuads             0.345907
SkiableTerrain_ac     0.314681
yearsOpen             0.294256
LongestRun_mi         0.291573
projectedDaysOpen     0.184608
TerrainParks          0.110023
double                0.048383

# **Model 5: Drop AdultWeekday Column**

In [57]:
df5 = pd.read_csv('step3_output.csv', index_col= None)

In [58]:
df5.head()

Name    state  ...  NightSkiing_ac  clusters
0       Alyeska Resort   Alaska  ...           550.0         1
1  Eaglecrest Ski Area   Alaska  ...             0.0         1
2     Hilltop Ski Area   Alaska  ...            30.0         1
3     Arizona Snowbowl  Arizona  ...             0.0         0
4  Sunrise Park Resort  Arizona  ...            80.0         0

[5 rows x 23 columns]

In [59]:
# Declare an explanatory variable, called x5, and assign it the result of dropping 'AdultWeekday','clusters','state','Name','summit_elev' and 'AdultWeekend' from the df5
x5 = df5.drop(['state','Name','AdultWeekend','summit_elev','clusters','AdultWeekday'], axis=1)
print(x5)

# Declare a response variable, called y5, and assign it the AdultWeekend column of the df5 
y5 = df5['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter x5 
scaler5 = preprocessing.StandardScaler().fit(x5)

# Declare a variable called x5_scaled, and assign it the result of calling the transform() method with parameter x5 
x5_scaled=scaler5.transform(x5)

     vertical_drop  fastQuads  ...  projectedDaysOpen  NightSkiing_ac
0             2500          2  ...         150.000000           550.0
1             1540          0  ...          90.000000             0.0
2              294          0  ...         152.000000            30.0
3             2300          0  ...         122.000000             0.0
4             1800          1  ...         104.000000            80.0
..             ...        ...  ...                ...             ...
325           1000          0  ...         120.053004             0.0
326            810          0  ...          77.000000             0.0
327           1571          0  ...         123.000000           110.0
328            990          0  ...         120.053004             0.0
329           1100          0  ...         120.053004             0.0

[330 rows x 17 columns]


In [60]:
# Get the 1-dimensional flattened array of our response variable y5 by calling the ravel() function on y5
y5 = y5.ravel()
# Call the train_test_split() function with the first two parameters set to x5_scaled and y5 
# Declare four variables, s5_train, s5_test, y5_train and y5_test separated by commas 
s5_train, s5_test, y5_train, y5_test = train_test_split(x5_scaled, y5, test_size=0.25, random_state=1)

In [61]:
lm5 = linear_model.LinearRegression()
model5 = lm5.fit(s5_train,y5_train)

In [62]:
# Make a variable called y5_pred and assign it the result of calling predict() on our model variable with parameter x5_test
y5_pred = model5.predict(s5_test)

In [63]:
explained_variance_score(y_true=y5_test, y_pred=y5_pred)

0.19530052423907873

In [64]:
mean_absolute_error(y_true=y5_test, y_pred=y5_pred)

15.594577238830501

In [65]:
lm5.intercept_

64.20764002530588

In [66]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_data5 = []
coef_data5 = pd.DataFrame(abs(lm5.coef_), x5.columns, columns=['Coefficient'])
print(coef_data5)

                   Coefficient
vertical_drop         7.048036
fastQuads            10.163979
quad                  7.826587
triple                8.139245
double                6.763739
surface               3.800667
total_chairs         17.829982
Runs                  1.456091
TerrainParks          0.881811
LongestRun_mi         0.896640
SkiableTerrain_ac     0.703635
Snow Making_ac        9.062039
daysOpenLastYear      1.319885
yearsOpen             0.689368
averageSnowfall       0.083391
projectedDaysOpen     1.628074
NightSkiing_ac        0.836592


In [67]:
coef_data5.sort_values(['Coefficient'], ascending=False)

Coefficient
total_chairs         17.829982
fastQuads            10.163979
Snow Making_ac        9.062039
triple                8.139245
quad                  7.826587
vertical_drop         7.048036
double                6.763739
surface               3.800667
projectedDaysOpen     1.628074
Runs                  1.456091
daysOpenLastYear      1.319885
LongestRun_mi         0.896640
TerrainParks          0.881811
NightSkiing_ac        0.836592
SkiableTerrain_ac     0.703635
yearsOpen             0.689368
averageSnowfall       0.083391

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.941 |  4.92 |'base_elev'|
| Model 2. | 0.930 |  5.34 |'state' base_elev'|
| Model 3. | 0.926 |  5.47 |'state' 'summit_elev' 'base_elev'|
| Model 4. | 0.925 |  5.54 |'state' 'summit_elev' 'base_elev' 'clusters'|
| Model 5. | 0.195 | 15.59 |'state' 'summit_elev' 'base_elev' 'clusters' 'AdultWeekday'|

Model Selection: Model 1